bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewShadeHandler.java

```diff
@@ -596,4 +611,100 @@ private void doRefreshShade(RefreshKind kind) {
+                .equals(c.getChannelTypeUID());
+
+        // current and required state of the vane channel
+        final boolean vaneChannelExisting = channelStream.anyMatch(vaneChannelFilter);
+        final boolean vaneChannelRequired = capabilities.usesVaneChannel();
+
+        final boolean secChannelExisting = channelStream.anyMatch(secChannelFilter);
+        final boolean secChannelRequired = capabilities.usesSecondaryChannel();
+
+        if ((vaneChannelExisting == vaneChannelRequired) && (secChannelExisting == secChannelRequired)) {
+            // nothing to do
+            return true;
+        }
+
+        // make a mutable copy of the original immutable channel list
+        channels = new ArrayList<>(channels);
+
+        boolean error = false;
+        if (vaneChannelRequired) {
+            // build the vane channel
+            Channel vaneChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(), HDPowerViewBindingConstants.CHANNEL_SHADE_VANE))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_VANE).withKind(ChannelKind.STATE)
+                    .withAcceptedItemType(CoreItemFactory.DIMMER)
+                    .withDescription(translater.getText("channel-type.hdpowerview.shade-vane.description"))
+                    .withLabel(translater.getText("channel-type.hdpowerview.shade-vane.label")).build();
+
+            // add the vane channel
+            error |= !channels.add(vaneChannel);
+        } else {
+            // remove the vane channel
+            error |= !channels.removeIf(vaneChannelFilter);
+        }
+
+        if (secChannelRequired) {
+            // build the secondary channel
+            Channel secChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(),
+                            HDPowerViewBindingConstants.CHANNEL_SHADE_SECONDARY_POSITION))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_SECONDARY).withKind(ChannelKind.STATE)
+                    .withAcceptedItemType(CoreItemFactory.DIMMER)
+                    .withDescription(translater.getText("channel-type.hdpowerview.shade-secondary.description"))
+                    .withLabel(translater.getText("channel-type.hdpowerview.shade-secondary.label")).build();
+
+            // add the secondary channel
+            error |= !channels.add(secChannel);
+        } else {
+            // remove the secondary channel
+            error |= !channels.removeIf(secChannelFilter);
+        }
+
+        if (error) {
+            // something went wrong
+            return false;
+        }
+
+        /*
+         * If we got this far then we will update the thing by building an identical copy of the target thing, except
+         * that it uses the newly modified channel list instead of the original one.
+         */
+        final Thing thing = editThing().withChannels(channels).build();
+        scheduler.submit(() -> updateThing(thing));
```

Comment:
Why do you need the scheduler for this?

Reply:
I just had a feeling that a class updating parts of itself should not be done directly within itself. But I think `poll()` is on a scheduled task so probably we can eliminate the scheduler here..

Reply:
Reopening as I still don't see any reason for this.

Reply:
Calling the framework's updateThing() method causes the framework to call the handler's thingUpdated() method, which in turn calls the handler's dispose() and initialize methods in turn. So in other words the handler instance is launching a series of actions that can have a significant impact on its own state. Therefore I feel more comfortable to definitively decouple any possible risk of the handler recursing method calls onto itself, by starting the avalanche separately from the task scheduler rather than from itself.

Reply:
Can you point me to the code which calls `dispose()`/`initialize()`? It doesn't sound right that the handler should need to be completely reinitialized because the model is updated. That would have high risk of becoming an infinite loop when `initialize()` would then again cause the model to be updated.

Reply:
Look into the BaseThingHandler code.

![Snag_195de5](https://user-images.githubusercontent.com/893994/185795496-b9f1f8da-742d-4f85-8444-fe594d835f5c.png)

 In updateThing() at line 456 it calls 'callback.thingUpdated()' and if you trace through the framework code, that eventually ends up calling 'BaseThingHandler.thingUpdate()' at line 152 ..

![Snag_1bb32f](https://user-images.githubusercontent.com/893994/185795592-5015d52f-16ec-474d-beba-f3daf7613466.png)

Reply:
> That would have high risk of becoming an infinite loop when initialize() would then again cause the model to be updated.

Exactly.  !!!

Reply:
`updateThing` (which you've shown above) calls this method in ThingHandlerCallback:

https://github.com/openhab/openhab-core/blob/8f745b0d3d5d263cffcd50b2a98989486af5415f/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java#L261-L289

This class implements `ThingManager`.

This is not the same method as in `ThingHandler`. So our `dispose()`/`initialize()` methods are not called.